src/solvers/petsc_nonlinear_solver.C

```diff
@@ -825,6 +870,41 @@ PetscNonlinearSolver<T>::solve (SparseMatrix<T> &  pre_in,  // System Preconditi
   // Should actually be a PetscReal, but I don't know which version of PETSc first introduced PetscReal
   Real final_residual_norm=0.;
 
+  // We don't want to do this twice because it resets
+  // SNESSetLagPreconditioner
+  if ((reuse_preconditioner()) && (!_setup_reuse))
+    {
+      _setup_reuse = true;
+      ierr = SNESSetLagPreconditionerPersists(_snes, PETSC_TRUE);
```

Comment:
Just a comment for myself as we may have already discussed this: `SNESSetLagPreconditionerPersists` was added in petsc/petsc@37ec4e1a76a069aca3c25363b0dc0a5311d937ea which dates back to PETSc 3.5, so I don't think any version guards are needed here.

Reply:
This I don't know.